cardano-testnet/src/Testnet/Components/Query.hs

```diff
@@ -364,65 +459,43 @@ getCurrentEpochNo epochStateView = withFrozenCallStack $ do
 -- or it becomes the same within the @maxWait@ epochs. If the value is not reached within the time frame,
 -- the test fails.
 assertNewEpochState
-  :: forall m era value.
-     (Show value, MonadAssertion m, MonadTest m, MonadIO m, Eq value, HasCallStack)
+  :: forall m era value. HasCallStack
+  => Show value
+  => Eq value
+  => MonadAssertion m
+  => MonadTest m
+  => MonadIO m
   => EpochStateView -- ^ Current epoch state view. It can be obtained using the 'getEpochStateView' function.
-  -> ConwayEraOnwards era -- ^ The ConwayEraOnwards witness for current era.
-  -> value -- ^ The expected value to check in the epoch state.
+  -> ShelleyBasedEra era -- ^ The ShelleyBasedEra witness for current era.
   -> EpochInterval -- ^ The maximum wait time in epochs.
-  -> Lens' (L.NewEpochState (ShelleyLedgerEra era)) value -- ^ The lens to access the specific value in the epoch state.
+  -> Lens' (L.NewEpochState (ShelleyLedgerEra era)) value
+  -- ^ The lens to access the specific value in the epoch state.
+  -> value -- ^ The expected value to check in the epoch state.
   -> m ()
-assertNewEpochState epochStateView ceo expected maxWait lens = withFrozenCallStack $ do
-  let sbe = conwayEraOnwardsToShelleyBasedEra ceo
-  mStateView <- watchEpochStateView epochStateView (checkEpochState sbe) maxWait
-  case mStateView of
-    Just () -> pure ()
-    Nothing -> do epochState <- getEpochState epochStateView
-                  val <- getFromEpochState sbe epochState
-                  if val == expected
-                  then pure ()
-                  else H.failMessage callStack $ unlines
-                         [ "assertNewEpochState: expected value not reached within the time frame."
-                         , "Expected value: " <> show expected
-                         , "Actual value: " <> show val
-                         ]
+assertNewEpochState epochStateView sbe maxWait lens expected = withFrozenCallStack $ do
+  mStateView <- watchEpochStateUpdate epochStateView maxWait (const checkEpochState)
+  when (isNothing mStateView) $ do
```

Comment:
`watchEpochStateUpdate` returns `Nothing` when the timeout has been exceeded (i.e the value of interest is not present in the ledger state). It's like we are checking "one last time here" which I'm not sure makes sense or is slightly misleading. Even more reason to not return a `Maybe` but instead a sum type that clearly says the value hasn't been found within the max wait period.

Reply:
Technically, the epoch state could change between `watchEpochStateUpdate` and `getFromEpochStateForEra`, so we could see different values. We're logging observed value so we need to call that function anyway - so doing check again won't hurt.

Reply:
>  Technically, the epoch state could change between watchEpochStateUpdate and getFromEpochStateForEra, so we could see different values. We're logging observed value so we need to call that function anyway - so doing check again won't hurt.

Add a comment making this clear.

Reply:
added